### ISSUE TITLE
query: API does not honor the per request timeout URL query parameter (one slow store makes all of thanos unavailable)

### ISSUE THREAD

**Issue by zecke** [thanos-io/thanos#3456]
**Thanos, Prometheus and Golang version used**:

Thanos v0.16.0
Go 1.15.3

**Object Storage Provider**:

S3

**What happened**:

Executing:

* thanos query --query.timeout=2m (default)
* curl http://127.1:9090/query_range?...&timeout=10s&partial_response=1

Result:
Query timesout after 2m and not 10s.

**What you expected to happen**:

I did expect the query to timeout after 10s (and deliver a partial response)

**How to reproduce it (as minimally and precisely as possible)**:

Make one thanos store really slow. We added a `srv.Context().Done()` into the Series() implementation of the Store. This should force timeouts and hopefully partial responses as other stores have data for the same time range from the same object store.

**Full logs to relevant components**:

The API correctly parses the timeout and creates the matching context:
https://github.com/thanos-io/thanos/blob/v0.16.0/pkg/api/query/v1.go#L232
https://github.com/thanos-io/thanos/blob/v0.16.0/pkg/api/query/v1.go#L336

But the querier prefers to use the global timeout over the per query timeout:

https://github.com/thanos-io/thanos/blob/v0.16.0/pkg/query/querier.go#L210-L211

This was changed in https://github.com/thanos-io/thanos/commit/c9c60c1b8c9e6cb8dd0c0a51513e3894c558b48d and seems a regression. After this change the q.selectTimeout is passed through the QueryableCreator.

The Prometheus Query Engine does seem to call q.Select in a blocking fashion and this means the entire query is blocked. Changing this to use q.ctx does fix the timeout handling but partial_responses seems to remain broken.

**Comment by bwplotka** [comment]
ack, sounds like a bug, thanks for reporting! :muscle:

**Comment by spaparaju** [comment]
Picking up this issue.

**Comment by zecke** [comment]
The issue #2144 is related to this problem for partial_response handling. Let's start with a request like the one below and have one _slow_ store (but we have N+2 stores for the same time range. This means from a user point of view we can tolerate some replicas not providing us with _any_ result).


```curl http://127.1:9090/query_range?...&timeout=10s&partial_response=1```

What will happen with the default thanos configuration (query timeout of 2m, store reply timeout of 0s being disable).

Without this issue fixed: The query will timeout after 2m with no data.
With this issue fixed: The query will timeout after 10s with no data.

But in both cases partial_responses remain broken. The default setting of store.reply_timeout and partial_responses are incompatible with each other. We can fix it by setting a store.reply_timeout but then it only works IFF the reply_timeout is smaller than the query timeout + time needed to aggregate the series.

With a per query timeout and partial_response (highly desired) handling the store reply timeout needs to be somehwat dynamic.

**Comment by spaparaju** [comment]
A single query (a request) can have only timeout value that it can honour. One way to solve 'partial_responses' (in the context of store-response-timeout and query-timeout) is  to 'fail the request' (with appropriate HTTP code) IF partial_responses flag is ON & query-timeout specified is less than the current underlying store-response-timeout. 
IMO dynamically adjusting timeouts at downstream components based on whatever user specify as query-timeout (which we can't control unless checks are in place) is not going to be trivial.

**Comment by zecke** [comment]
Thanks! Where is the best way to talk about a different approach for the query component. It seems like a bigger change. I have some ideas but there are many open questions.

**Comment by zecke** [comment]
I think this is still open.

**Comment by spaparaju** [comment]
Yes it is open. Here is the [PR](https://github.com/thanos-io/thanos/pull/3539).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [thanos-io/thanos] query: API d..." ]
>Yes it is open. Here is the PR.

can you teach 'stalebot' to not close tickets if there is an open PR references the issue?

**Comment by yeya24** [comment]
We can add exempt labels to the issues so that the bot can exclude them.

**Comment by yeya24** [comment]
Still an issue for us. And I am trying to understand the flag `store.response-timeout` more.
It means the timeout of receiving one series from a store, not receiving all serieses from a store during a query lifecycle if I understand correctly?
